File: packages/telemetry/src/query/parse-response.ts
export type ParseResult =
	| { ok: true; value: number }
	| { ok: false; reason: 'no_data' | 'non_finite_value' };

type InstantQueryData = {
	resultType: string;
	result: unknown;
};

function parseSampleValue(valueString: string): ParseResult {
	if (valueString === 'NaN' || valueString === '+Inf' || valueString === '-Inf') {
		return { ok: false, reason: 'non_finite_value' };
	}
	const value = Number.parseFloat(valueString);
	if (!Number.isFinite(value)) {
		return { ok: false, reason: 'non_finite_value' };
	}
	return { ok: true, value };
}

function parseVectorResult(result: unknown): ParseResult {
	if (!Array.isArray(result) || result.length === 0) {
		return { ok: false, reason: 'no_data' };
	}
	const first = result[0] as { value?: [number, string] };
	if (!first?.value || first.value.length < 2) {
		return { ok: false, reason: 'no_data' };
	}
	return parseSampleValue(first.value[1]);
}

function parseScalarResult(result: unknown): ParseResult {
	if (!Array.isArray(result) || result.length < 2) {
		return { ok: false, reason: 'no_data' };
	}
	const valueString = result[1];
	if (typeof valueString !== 'string') {
		return { ok: false, reason: 'no_data' };
	}
	return parseSampleValue(valueString);
}

export function parseInstantQueryResult(data: InstantQueryData): ParseResult {
	if (data.resultType === 'vector') {
		return parseVectorResult(data.result);
	}
	if (data.resultType === 'scalar') {
		return parseScalarResult(data.result);
	}
	return { ok: false, reason: 'no_data' };
}

File: kotlin/modules/telemetry/src/main/kotlin/com/ffpromo/telemetry/query/ParseResponse.kt
package com.ffpromo.telemetry.query

import com.ffpromo.telemetry.client.PrometheusInstantQueryData
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive

sealed class ParseResult {
    data class Ok(val value: Double) : ParseResult()
    data class Fail(val reason: ParseFailReason) : ParseResult()
}

enum class ParseFailReason {
    no_data,
    non_finite_value,
}

private fun parseSampleValue(valueString: String): ParseResult {
    if (valueString == "NaN" || valueString == "+Inf" || valueString == "-Inf") {
        return ParseResult.Fail(ParseFailReason.non_finite_value)
    }
    val value = valueString.toDoubleOrNull()
    if (value == null || !value.isFinite()) {
        return ParseResult.Fail(ParseFailReason.non_finite_value)
    }
    return ParseResult.Ok(value)
}

private fun parseVectorResult(result: JsonElement): ParseResult {
    val array = result as? JsonArray ?: return ParseResult.Fail(ParseFailReason.no_data)
    if (array.isEmpty()) {
        return ParseResult.Fail(ParseFailReason.no_data)
    }
    val first = array.first() as? JsonObject ?: return ParseResult.Fail(ParseFailReason.no_data)
    val valueTuple = first["value"]?.jsonArray ?: return ParseResult.Fail(ParseFailReason.no_data)
    if (valueTuple.size < 2) {
        return ParseResult.Fail(ParseFailReason.no_data)
    }
    return parseSampleValue(valueTuple[1].jsonPrimitive.content)
}

private fun parseScalarResult(result: JsonElement): ParseResult {
    val array = result as? JsonArray ?: return ParseResult.Fail(ParseFailReason.no_data)
    if (array.size < 2) {
        return ParseResult.Fail(ParseFailReason.no_data)
    }
    return parseSampleValue(array[1].jsonPrimitive.content)
}

fun parseInstantQueryResult(data: PrometheusInstantQueryData): ParseResult =
    when (data.resultType) {
        "vector" -> parseVectorResult(data.result)
        "scalar" -> parseScalarResult(data.result)
        else -> ParseResult.Fail(ParseFailReason.no_data)
    }
